compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/LimitScatterBounds.cpp

```diff
@@ -0,0 +1,281 @@
+            .getResult(0);
+  }
+
+  if (cast<ShapedType>(inbounds.getType()).getRank() < update0Ty.getRank()) {
+    llvm::SmallVector<bool> isBatchDim(update0Ty.getRank(), true);
+    for (auto dim : updateWindowDims) {
+      isBatchDim[dim] = false;
+    }
+
+    llvm::SmallVector<AffineExpr> inExprs;
+    for (int i = 0; i < isBatchDim.size(); i++) {
+      if (isBatchDim[i]) {
+        inExprs.push_back(b.getAffineDimExpr(i));
+      }
+    }
+
+    ShapedType outTy = update0Ty.clone(b.getI1Type());
+    AffineMap inMap =
+        AffineMap::get(outTy.getRank(), 0, inExprs, b.getContext());
+    AffineMap outMap = b.getMultiDimIdentityMap(outTy.getRank());
+
+    SmallVector<utils::IteratorType> iters(outTy.getRank(),
+                                           utils::IteratorType::parallel);
+
+    Value empty =
+        b.create<tensor::EmptyOp>(outTy.getShape(), outTy.getElementType());
+    inbounds =
+        b.create<linalg::GenericOp>(
+             TypeRange{empty.getType()}, ValueRange{inbounds},
+             ValueRange{empty}, ArrayRef<AffineMap>{inMap, outMap}, iters,
+             [](OpBuilder &bb, Location loc, ValueRange args) {
+               bb.create<linalg::YieldOp>(loc, args[0]);
+             })
+            .getResult(0);
+  }
+
+  llvm::SmallVector<Value> newInputs(op.getInputs());
+
+  ShapedType inputTy = cast<ShapedType>(newInputs.front().getType());
+  Value emptyBool =
+      b.create<tensor::EmptyOp>(inputTy.getShape(), b.getI1Type());
+  newInputs.push_back(emptyBool);
+
+  llvm::SmallVector<Value> newUpdates(op.getUpdates());
+  newUpdates.push_back(inbounds);
+
+  llvm::SmallVector<Type> newResultTys(op.getResultTypes());
+  newResultTys.push_back(emptyBool.getType());
+
+  auto scatter = b.create<mlir::stablehlo::ScatterOp>(
+      newResultTys, newInputs, bounded, newUpdates, dimnums,
+      op.getIndicesAreSorted(), op.getUniqueIndices());
+
+  // Replace the existing yield with conditional results and the boolean update:
+  for (int i = 0; i < op.getNumResults(); ++i) {
+    op.getResult(i).replaceAllUsesWith(scatter.getResult(i));
+  }
+
+  // Clone the region into the new scatter and insert the new boolean udpates:
+  Region &srcRegion = op.getUpdateComputation();
+  Region &dstRegion = scatter.getUpdateComputation();
+
+  IRMapping mapping;
+  srcRegion.cloneInto(&dstRegion, mapping);
+
+  const int64_t numUpdates = op.getUpdates().size();
+  dstRegion.front().insertArgument(
+      numUpdates * 2, RankedTensorType::get({}, b.getI1Type()), op.getLoc());
+  dstRegion.front().insertArgument(
+      numUpdates, RankedTensorType::get({}, b.getI1Type()), op.getLoc());
+
+  // Update the return to only conditionally return the new value:
+  Operation *terminator = dstRegion.front().getTerminator();
+  b.setInsertionPoint(terminator);
+
+  Value argInbounds = dstRegion.front().getArguments().back();
+  llvm::SmallVector<Value> terminatorOperands;
+  for (int i = 0; i < terminator->getNumOperands(); ++i) {
```

Comment:
Is this verified to match by here? Could we do a zip here instead?

Reply:
Sadly the are not guaranteed to have the same number. In fact we are guaranteed to have more block args. The first `N` block args match to the read values for the `N` destinations. So we could slice and zip but that felt more complicated than purely selecting the indexing.